### PR TITLE
Implement zoom using scaled SDL_RenderCopy.

### PR DESCRIPTION
**Pull request by flackr**
This allows removing the zoom texture allocations and removing an intermediate draw to that texture before drawing to the screen noticeably increasing the frame rate which may address issues such as #1420.

### CODE REVIEW

**Comment by tobylane** [comment]
You can ignore this run's error, it's not your changes. I locally generated your PR's documentation fine.

**Comment by TheCycoONE** [comment]
Rebasing on master will fix the pipeline. The change looks good to me. Will test. Thanks @flackr

**Comment by TheCycoONE** [comment]
This PR prevents the game from hanging when zooming out on a computer here. Escalating to 0.65.

**Comment by TheCycoONE** [comment]
There is a regression: on windows you can set the resolution lower than native and then press alt+enter to full screen, which scales the full UI. There is a noticeable reduction in the quality of the result.

**Comment by flackr** [comment]
I'm a bit confused, is that different than regular zooming out? Or is this a different part of the UI which looks bad? Could you share a screenshot?

I know that `raw_bitmap::draw` has a separate scaling which for best quality we should really be passing forward into `render_target::draw`. I'm not quite sure where raw bitmaps are used or when we set a scaling on them but I can roll that into this change which should eliminate possible additional artifacts there.

**Comment by TheCycoONE** [comment]
No, not zoom. In windows you can set a lower screen resolution and then switch to full screen and SDL applies a fake full screen by scaling everything on the screen. (Same effect as SDL_RenderSetLogicalSize)

After Patch:
windows_scale

Before:
windows_scale_old

File: CorsixTH/Src/th_gfx_sdl.cpp [hunk 1]
```diff
@@ -347,37 +369,12 @@ void render_target::destroy() {
 }
 
 bool render_target::set_scale_factor(double fScale, scaled_items eWhatToScale) {
-  flush_zoom_buffer();
   scale_bitmaps = false;
 
   if (fScale <= 0.000) {
     return false;
   } else if (eWhatToScale == scaled_items::all && supports_target_textures) {
```
Comment (TheCycoONE) [inline CorsixTH/Src/th_gfx_sdl.cpp:376]: supports_target_textures should be unnecessary now

**Comment by TheCycoONE** [comment]
Additionally there are artifacts in the shadows at some zoom levels:
shadow_artifact

**Comment by TheCycoONE** [comment]
https://github.com/CorsixTH/CorsixTH/blob/ac4f26df9e54e778cfea257a1f5ba06538f28900/CorsixTH/Src/th_map.cpp#L1008 is where the shadows are rendered, just above is the tiles.

**Comment by TheCycoONE** [comment]
There is a class of hardware that this patch benefits as is. A cop out, but perhaps it should be a (preferably runtime) flag for now?

**Comment by flackr** [comment]
Sounds reasonable to me. Then we can incrementally fix the visual artifacts and get it working well on all platforms, but it can benefit users already :-).

Do you have a system for runtime flags / reference to some code I can copy for that?

**Comment by TheCycoONE** [comment]
We do: https://github.com/CorsixTH/CorsixTH/blob/master/CorsixTH/Lua/config_finder.lua
I can help with that aspect.

**Comment by TheCycoONE** [comment]
I've pushed the code for setting the runtime option (back to the render_target). Now it's a matter of restoring the old behaviour conditionally. If you're willing, I'll leave that part up to you.

**Comment by flackr** [comment]
Sounds good, I should have time this afternoon.

**Comment by flackr** [comment]
I believe this is now properly behind a flag with the zoom texture behavior without it set.

**Review by TheCycoONE — CHANGES_REQUESTED** [review]
This toggle is in a bit of a limbo as to whether it would support an in game toggle at the moment.

To do so:
- The SDL_SetHint for linear mode in `create` should be moved to an else condition of `if (direct_zoom)` in the `update` function of th_gfx_sdl.cpp.
- global_scale_factor should be reset to 1.0
- All textures would need to be flushed / recreated after the hint is changed. The documentation says that `SDL_HINT_RENDER_SCALE_QUALITY` is applied to a texture at creation rather than at copy (which explains why I didn't see any effect toggling it in set_scale_factor)

If we don't want it to be an in game toggle then this block belongs in the `create` function, either way we only want to set SDL_HINT_RENDER_SCALE_QUALITY in one place (for now - as the feature evolves I suspect we might assign different scale modes to different textures).

File: CorsixTH/Src/th_gfx_sdl.cpp [hunk 2]
```diff
@@ -95,6 +95,30 @@ uint8_t convert_6bit_to_8bit_colour_component(uint8_t colour_component) {
...
 }
 
+//! Get the enclosing rect of an SDL_Rect scaled by the given scale factor.
+//  Scaling the rect is location dependent, as non-integer scale factors
+//  require that same size rects be +/-1 pixel different in size depending
+//  on location.
+/*!
+    @param rect Pointer to the SDL_Rect to be scaled.
+    @param scale_factor Scale to be applied to the rectangle.
+    @return Enclosing SDL_Rect of the input rect scaled by scale_factor.
```
[the remark this answers is not in the document]
Comment (flackr) [reply CorsixTH/Src/th_gfx_sdl.cpp]: Oops, that was leftover from when it was the return value. Fixed.

**Comment by flackr** [comment]
> This toggle is in a bit of a limbo as to whether it would support an in game toggle at the moment.
> 
> To do so:
> 
> * The SDL_SetHint for linear mode in `create` should be moved to an else condition of `if (direct_zoom)` in the `update` function of th_gfx_sdl.cpp.
> * global_scale_factor should be reset to 1.0
> * All textures would need to be flushed / recreated after the hint is changed. The documentation says that `SDL_HINT_RENDER_SCALE_QUALITY` is applied to a texture at creation rather than at copy (which explains why I didn't see any effect toggling it in set_scale_factor)
> 
> If we don't want it to be an in game toggle then this block belongs in the `create` function, either way we only want to set SDL_HINT_RENDER_SCALE_QUALITY in one place (for now - as the feature evolves I suspect we might assign different scale modes to different textures).

Sounds good. I've moved the flag to the create function as suggested. I'm assuming create only happens once so we should have zoom_texture == nullptr and global_scale_factor == 1.0 right? If not, I can reset them appropriately.

**Review by TheCycoONE — APPROVED** [review]

**Comment by TheCycoONE** [comment]
> > This toggle is in a bit of a limbo as to whether it would support an in game toggle at the moment.
> > To do so:
> > 
> > * The SDL_SetHint for linear mode in `create` should be moved to an else condition of `if (direct_zoom)` in the `update` function of th_gfx_sdl.cpp.
> > * global_scale_factor should be reset to 1.0
> > * All textures would need to be flushed / recreated after the hint is changed. The documentation says that `SDL_HINT_RENDER_SCALE_QUALITY` is applied to a texture at creation rather than at copy (which explains why I didn't see any effect toggling it in set_scale_factor)
> > 
> > If we don't want it to be an in game toggle then this block belongs in the `create` function, either way we only want to set SDL_HINT_RENDER_SCALE_QUALITY in one place (for now - as the feature evolves I suspect we might assign different scale modes to different textures).
> 
> Sounds good. I've moved the flag to the create function as suggested. I'm assuming create only happens once so we should have zoom_texture == nullptr and global_scale_factor == 1.0 right? If not, I can reset them appropriately.

For completeness - yes it's only called once, immediately after the constructor.

**Comment by TheCycoONE** [comment]
Since we are delaying, and this PR has been tested, I'm merging it in.

**Comment by flackr** [comment]
Now that this is merged shall I file issues for the artifacts that would block making this the default rendering path?

**Comment by TheCycoONE** [comment]
Yes please

**Comment by flackr** [comment]
Done, filed #1929 and #1930 along with a rough idea of how I think they can be fixed.

**Pull request merged**